tokio/src/sync/mpsc/unbounded.rs

```diff
@@ -172,6 +172,56 @@ impl<T> UnboundedReceiver<T> {
         poll_fn(|cx| self.poll_recv(cx)).await
     }
 
+    /// Receives the next values for this receiver and extends `buffer`.
+    ///
+    /// This method returns the number of values populated in `buffer`.
+    ///
+    /// This method returns 0 if the channel has been closed and there are
+    /// no remaining messages in the channel's queue. This indicates that no
+    /// further values can ever be received from this `Receiver`. The channel is
+    /// closed when all senders have been dropped, or when [`close`] is called.
+    ///
+    /// If there are no messages in the channel's queue, but the channel has
+    /// not yet been closed, this method will sleep until a message is sent or
+    /// the channel is closed.
+    ///
+    /// If at the time of the call `buffer` has no unused capacity,
+    /// additional elements are reserved. Otherwise `recv_many` populates
+    /// `buffer` with no more elements than its unused capacity.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe. If `recv` is used as the event in a
+    /// [`tokio::select!`](crate::select) statement and some other branch
+    /// completes first, it is guaranteed that no messages were received on this
+    /// channel.
+    ///
+    /// [`close`]: Self::close
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = mpsc::unbounded_channel();
+    ///
+    ///     tokio::spawn(async move {
+    ///         tx.send("hello").unwrap();
+    ///     });
+    ///
+    ///     let mut buffer : Vec<&str> = Vec::with_capacity(5);
+    ///     assert_eq!(1, rx.recv_many(&mut buffer).await);
+    ///     assert_eq!(vec!["hello"], buffer);
+    ///     assert_eq!(0, rx.recv_many(&mut buffer).await);
+    /// }
+    /// ```
```

Comment:
The examples are very different. Perhaps we could use the same example? I imagine a good example could work like this:

 * First send two messages, then receive both in one go.
 * Then spawn a task that sends a message, and receive it with `recv_many`. Include a comment saying that this call is guaranteed to not return 0 since the channel is not closed.
 * Drop the last sender.
 * Then call `recv_many` again and get 0 because the channel is closed.

Reply:
I've set it up so that all senders are dropped to close the channel.
Two paths not covered:

- Demonstrate that the buffer's unused capacity is a bound on the number of elements returned.  
- Close the channel but have outstanding permits/messages; similar to the first doc example for `recv`

Reply:
I've added two new tests analogous to the doctests to sync_mpsc.rs to cover these behaviors explicitly.